### PR TITLE
Add INDI Rotator support

### PR DESCRIPTION
**Pull request by comptonizing**
Good day,

First of all, thank you for your great work.

I took the liberty of adding support for INDI rotators. Large portions of this code have been adapted from the INDI scope and camera implementation.

I am aware of #1089 but am not sure how to proceed with regards to it. In case this PR is useful and being considered for merge, maybe some comments from the maintainers could help. Of course I'd be very happy with someone else adapting it to INDI 2, in line with the changes by Jasem.

Would address #1041 

Thank you,
Philipp

### CODE REVIEW

**Review by d33psky — APPROVED** [review]
Nice work !
Looks fine to me.

File: rotator_indi.cpp
Comment (agalasso) [inline rotator_indi.cpp:17]: Some of the older source files that came from the original Stark Labs PHD (PHD1) contained this copyright referencing to Stark Labs.   In new code that is not from Stark Labs we no longer reference Stark labs in the copyright notice (though in some cases like this the copyright got copy-pasted inadvertently)
Our preference would be to use the copyright notice referring only to openphdguiding.org -- you can copy the text from scope_manual_pointing.h.

File: rotator_indi.h
Comment (agalasso) [inline rotator_indi.h:17]: ditto

**Review by agalasso — CHANGES_REQUESTED** [review]
@comptonizing  thanks for the PR, nice work!
Could you please address the comments, then we'll do another round of review and should be able to get this merged.

**Comment by agalasso** [comment]
@comptonizing I wanted to also mention that #1089 has merged to master, so we'll need to make sure this code is in sync with those changes, thanks!

**Comment by comptonizing** [comment]
@agalasso I think I resolved the things pointed out above, thanks for the comments.
I also saw that the merge of the new INDI version was reverted, so my changes are currently in sync with the master branch. I'm not sure what Jasem's plans are for migrating the entire phd2 project to the new INDI version, so I would like to leave this PR at your discretion.

**Comment by agalasso** [comment]
@comptonizing it took us a while to get here, but once #1129 merges that should un-block this pr.  I plan to merge #1129 today or tomorrow.

**Comment by deletio** [comment]
Greetings everyone. It has been a while without updates here, so I gave it a try myself.

I locally merged the source branch into the current master and used #1129 as reference to refactor rotator_indi.cpp.
The build succeeds and I am able to connect from PHD2 to my Wanderer Rotator Mini V2. 

image

Attaching the log with a quick connection/disconnection of the rotator for your review.
[PHD2_DebugLog_2024-03-05_015115.txt](https://github.com/OpenPHDGuiding/phd2/files/14484649/PHD2_DebugLog_2024-03-05_015115.txt)

Please let me know if there are any specific test cases you would like me to run.
Otherwise please advise whether I should try to commit into @comptonizing's branch to enhance this current PR or raise my own. I would rather keep this PR, since it has a good history, but will need write access to the source branch.

**Comment by comptonizing** [comment]
> Greetings everyone. It has been a while without updates here, so I gave it a try myself.
> 
> I locally merged the source branch into the current master and used #1129 as reference to refactor rotator_indi.cpp. The build succeeds and I am able to connect from PHD2 to my Wanderer Rotator Mini V2.
> image
> 
> Attaching the log with a quick connection/disconnection of the rotator for your review. [PHD2_DebugLog_2024-03-05_015115.txt](https://github.com/OpenPHDGuiding/phd2/files/14484649/PHD2_DebugLog_2024-03-05_015115.txt)
> 
> Please let me know if there are any specific test cases you would like me to run. Otherwise please advise whether I should try to commit into @comptonizing's branch to enhance this current PR or raise my own. I would rather keep this PR, since it has a good history, but will need write access to the source branch.
> 
> **upd**. Uploaded to a fresh fork, the changes are here: [deletio@e674291](https://github.com/deletio/phd2-indi-rotator/commit/e674291c946c15033ca577e28253e1859dc0cb25)

Thanks a lot for this! For a while I couldn't rebase against master because I always ran into segfaults and wanted to work on this again this week, but if you've done the work already there really isn't any reason to do so again.
I granted you access to my repository if that helps in any way.

Of course the last word lies with @agalasso

File: rotator.cpp
Comment (agalasso) [inline rotator.cpp:96]: nit: please setup your editor to use spaces and not hard tabs and use 4-space indentation
```suggestion
            rotator = INDIRotatorFactory::MakeINDIRotator();
        }
```

File: rotators.h
Comment (agalasso) [inline rotators.h:50]: let's just define ROTATOR_INDI unconditionally -- it's available on all three platforms PHD2 runs on

**Review by agalasso — COMMENTED** [review]
looks good!

please address the suggestions, and we should be able to merge it

**Comment by deletio** [comment]
> > Greetings everyone. It has been a while without updates here, so I gave it a try myself.
> > I locally merged the source branch into the current master and used #1129 as reference to refactor rotator_indi.cpp. The build succeeds and I am able to connect from PHD2 to my Wanderer Rotator Mini V2.
> > image
> > Attaching the log with a quick connection/disconnection of the rotator for your review. [PHD2_DebugLog_2024-03-05_015115.txt](https://github.com/OpenPHDGuiding/phd2/files/14484649/PHD2_DebugLog_2024-03-05_015115.txt)
> > Please let me know if there are any specific test cases you would like me to run. Otherwise please advise whether I should try to commit into @comptonizing's branch to enhance this current PR or raise my own. I would rather keep this PR, since it has a good history, but will need write access to the source branch.
> > **upd**. Uploaded to a fresh fork, the changes are here: [deletio@e674291](https://github.com/deletio/phd2-indi-rotator/commit/e674291c946c15033ca577e28253e1859dc0cb25)
> 
> Thanks a lot for this! For a while I couldn't rebase against master because I always ran into segfaults and wanted to work on this again this week, but if you've done the work already there really isn't any reason to do so again. I granted you access to my repository if that helps in any way.
> 
> Of course the last word lies with @agalasso

I was hesitant to force-push the rebase, so for now I have pushed it as a new branch https://github.com/comptonizing/phd2/tree/indi-rotator-refactored. Please see if it looks good to you and maybe check if it still segfaults as your previous attempts did. Will happily force-push onto the current branch (indi-rotator) at your first signal.

@agalasso's comments have been addressed in the last commit.

File: config_indi.h
Comment (agalasso) [inline config_indi.h:56]: minor, to simplify future diffs:
```suggestion
    INDI_TYPE_ROTATOR,
```

**Comment by deletio** [comment]
I ended up updating the source branch of this PR, so it is now in the final state with all comments addressed.

**Comment by comptonizing** [comment]
> I ended up updating the source branch of this PR, so it is now in the final state with all comments addressed.

Thank you, seems to be working fine, with one minor problem: The rotator angle in PHD2 is shown as UKNOWN until the rotator has actually changed position once. This is problematic because it forces a recalibration upon this first rotator movement. It appears as updating the angle upon connecting to the driver is not working.

**Comment by deletio** [comment]
> > I ended up updating the source branch of this PR, so it is now in the final state with all comments addressed.
> 
> Thank you, seems to be working fine, with one minor problem: The rotator angle in PHD2 is shown as UKNOWN until the rotator has actually changed position once. This is problematic because it forces a recalibration upon this first rotator movement. It appears as updating the angle upon connecting to the driver is not working.

I also see that in the log as
`GetDouble("/profile/1/scope/calibration/rotatorAngle", -888.000000) returns -888.000000`

looking into this

**Comment by comptonizing** [comment]
> > Thank you, seems to be working fine, with one minor problem: The rotator angle in PHD2 is shown as UKNOWN until the rotator has actually changed position once. This is problematic because it forces a recalibration upon this first rotator movement. It appears as updating the angle upon connecting to the driver is not working.
> 
> I also see that in the log as `GetDouble("/profile/1/scope/calibration/rotatorAngle", -888.000000) returns -888.000000`
> 
> looking into this

I think I was working correctly with my previous version before the introduction of INDI 2.0 if that helps in any way

**Comment by deletio** [comment]
> > I also see that in the log as `GetDouble("/profile/1/scope/calibration/rotatorAngle", -888.000000) returns -888.000000`
> > looking into this
> 
> I think I was working correctly with my previous version before the introduction of INDI 2.0 if that helps in any way

Of course it helps 👍 

The log line I observed was actually not related: it was merely indicating that the calibration being restored did not have a rotator angle in it.

But indeed the angle was not updated on connection, which should be fixed now. Would be great if you could confirm, because I am not sure where exactly you were observing the UNKNOWN angle.

Also, I think I have an idea what could be causing segfaults. Calling `RotatorINDI::updateAngle()` without prior initialization of the `angle_prop` pointer could do that.

**Comment by comptonizing** [comment]
> > > I also see that in the log as `GetDouble("/profile/1/scope/calibration/rotatorAngle", -888.000000) returns -888.000000`
> > > looking into this
> > 
> > 
> > I think I was working correctly with my previous version before the introduction of INDI 2.0 if that helps in any way
> 
> Of course it helps 👍
> 
> The log line I observed was actually not related: it was merely indicating that the calibration being restored did not have a rotator angle in it.
> 
> But indeed the angle was not updated on connection, which should be fixed now. Would be great if you could confirm, because I am not sure where exactly you were observing the UNKNOWN angle.

The "UNKNOWN" showed up in the statistics pane of the GUI. I just tested your changes, they're working perfectly!
> 
> Also, I think I have an idea what could be causing segfaults. Calling `RotatorINDI::updateAngle()` without prior initialization of the `angle_prop` pointer could do that.

Sorry, it appears I didn't explain properly. After the INDI 2.0 stuff was merged, the master branch itself got me segfaults, at least here on Ubuntu 22.04 with INDI 2.0.6. It was completely unrelated to my changes and I couldn't find the problem quickly.

(If you want to you can add your name to the copyright section of the new files)

**Comment by deletio** [comment]
Will do shortly, thank you!

**Comment by deletio** [comment]
@comptonizing Since you drew my attention to the guide stats pane which I never used before, here is one more minor update that makes sure the rotator angle line updates not only when the mount is [dis]connected, but also when the rotator is [dis]connected or actually rotated.

**Comment by comptonizing** [comment]
> @comptonizing Since you drew my attention to the guide stats pane which I never used before, here is one more minor update that makes sure the rotator angle line updates not only when the mount is [dis]connected, but also when the rotator is [dis]connected or actually rotated.

Awesome, thanks! I wanted to do this initially but since the status updated during calibration and guiding anyway I didn't deem it necessary. It's pretty nice now, you get immediate feedback

**Comment by agalasso** [comment]
> (If you want to you can add your name to the copyright section of the new files)

re: copyright: if you do not mind, I'd prefer to say:
> Copyright (C) 2024 openphdguiding.org

Having individual developers copyright individual source files was a pattern that got established early on and I would like to get away from it. It is unwieldy due to the BSD license 2d clause requirement that all the copyright statements be reproduced.

I would prefer give both of you credit by adding your names to the contributors list (Help > About) as a followup PR after this is merged, and not have to add an additional copyright statement as required by the 2d clause of the license.

File: rotator_indi.cpp
Comment (agalasso) [inline rotator_indi.cpp:7]: https://github.com/OpenPHDGuiding/phd2/pull/1091#issuecomment-1979760396
```suggestion
*  Created by Philipp Weber and Kirill M. Skorobogatov
*  Copyright (c) 2024 openphdguiding.org
```

**Comment by deletio** [comment]
Updated rotator_indi.* as requested. @agalasso should we also update files with one author?

**Comment by deletio** [comment]
One more consideration: should we adapt the connection method from #1177 to rotators as well?

It seems something similar was actually in place originally, but removed with INDI2 updates (and indeed I cleaned up that code myself as part of refactoring here).

The current state of the code has worked smoothly for me, so I am not sure of the merit to change it. Whereas #1177 clearly addressed connectivity issues stated therein. What do you think?

**Review by agalasso — APPROVED** [review]
@comptonizing is this ready to merge?

**Comment by comptonizing** [comment]
> @comptonizing is this ready to merge?

Yes, I had a look, tested it again, and everything seems to work like a charm.

Sorry, I got sick last week.

**Pull request merged**
